modules/binary-tuple/src/main/java/org/apache/ignite/internal/binarytuple/inlineschema/TupleWithSchemaMarshalling.java

```diff
@@ -0,0 +1,378 @@
+        if (raw == null) {
+            return null;
+        }
+        if (raw.length < 8) {
+            throw new UnmarshallingException("byte[] length can not be less than 8");
+        }
+
+        // Read first int32.
+        ByteBuffer buff = ByteBuffer.wrap(raw).order(BYTE_ORDER);
+        int size = buff.getInt(0);
+        if (size < 0) {
+            throw new UnmarshallingException("Size of the tuple can not be less than zero");
+        }
+
+        // Read second int32.
+        int valueOffset = buff.getInt(4);
+        if (valueOffset < 0) {
+            throw new UnmarshallingException("valueOffset can not be less than zero");
+        }
+        if (valueOffset > raw.length) {
+            throw new UnmarshallingException(
+                    "valueOffset can not be greater than byte[] length, valueOffset: "
+                            + valueOffset + ", length: " + raw.length
+            );
+        }
+
+        ByteBuffer schemaBuff = buff
+                .position(8).limit(valueOffset)
+                .slice().order(BYTE_ORDER);
+        ByteBuffer valueBuff = buff
+                .position(valueOffset).limit(raw.length)
+                .slice().order(BYTE_ORDER);
+
+        BinaryTupleReader schemaReader = new BinaryTupleReader(size * 2, schemaBuff);
+        BinaryTupleReader valueReader = new BinaryTupleReader(size, valueBuff);
+
+        String[] columns = new String[size];
+        ColumnType[] types = new ColumnType[size];
+        Tuple tup = Tuple.create(size);
+
+        int readerInd = 0;
+        int i = 0;
+        while (i < size) {
+            String colName = schemaReader.stringValue(readerInd++);
+            int colTypeId = schemaReader.intValue(readerInd++);
+
+            columns[i] = colName;
+            types[i] = ColumnType.getById(colTypeId);
+
+            i += 1;
+        }
+
+        int k = 0;
+        while (k < size) {
+            setColumnValue(valueReader, tup, columns[k], types[k].id(), k);
+            k += 1;
+        }
+
+        return tup;
+    }
+
+    private static BinaryTupleBuilder schemaBuilder(String[] columns, ColumnType[] types) {
+        BinaryTupleBuilder builder = new BinaryTupleBuilder(columns.length * 2, columns.length * 2);
+
+        for (int i = 0; i < columns.length; i++) {
+            builder.appendString(columns[i]);
+            builder.appendInt(types[i].id());
+        }
+
+        return builder;
+    }
+
+    private static BinaryTupleBuilder valueBuilder(String[] columnNames, ColumnType[] types, Object[] values) {
+        BinaryTupleBuilder builder = new BinaryTupleBuilder(values.length, values.length);
+
+        for (int i = 0; i < values.length; i++) {
+            ColumnType type = types[i];
+            Object v = values[i];
+
+            appender(type, columnNames[i]).accept(builder, v);
+        }
+
+        return builder;
+    }
+
+    private static BiConsumer<BinaryTupleBuilder, Object> appender(ColumnType type, String name) {
```

Comment:
This method returns a `BiConsumer` that is immediately invoked, which causes unnecessary allocation and indirection.

We can either inline this method, or make it `void` and perform the append action immediately.